keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,527 @@
+
+Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Proposed default fields:
+* `ts` - timestamp as Unix time (required, `float64`)
+* `v` - verbosity (required, `int`, default 4)
+* `err` - error string (optional, `string`)
+* `msg` - message (required, `string`)
+
+Default Serialization:
+* primitive types (`int`, `float`) - Compatible with [golang json package](https://golang.org/pkg/encoding/json/#Marshal)
+* namespaced Kubernetes objects (`Pod`, `Deployment`) - JSON object consisting of `name` and `namespace`. Example `{"name": "kubedns", "namespace": "kube-system"}`
+* not namespaced Kubernetes objects (`Node`) - JSON object consisting of `name`. Example `{"name": "cluster1-vm-72x33b8p-34jz"}`
+* arbitrary structure (`http.Request`) - JSON objects consisting of public non-pointer fields.
+
+Proposed serialization of Kubernetes objects will not include additional fields beside those proposed available in `text` format.
+Discussion for additional fields needed in representation of Kubernetes objects is out of scope of this KEP.
+
+Serialization of arbitrary structures will ignore pointers to avoid hard to predict performance hits caused by deep nested structures or cycles.
+
+Provided serialization should function as good default. Caller can customize how objects is serialized by changing struct beforehand.
+
+Example:
+
+```go
+klog.Infof("Updated pod %s status to ready", pod.name)
+```
+That would result in log (pretty printed)
+
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "msg": "Updated pod kubedns status to ready"
+}
+```
+
+And
+
+```go
+klog.ErrorS(err, "Failed to update pod status")
+```
+
+That would result in log (pretty printed)
+
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "err": "timeout",
+   "msg": "Failed to update pod status"
+}
+```
+
+And
+
+```go
+pod := corev1.Pod{Name: "kubedns", Namespace: "kube-system", ...}
+klog.InfoS("Pod status updated", "pod", pod, "status", "ready")
+```
+
+That would result in log (pretty printed)
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "msg": "Pod status updated",
+   "pod":{
+      "name": "nginx-1",
+      "namespace": "default"
+   },
+   "status": "ready"
+}
+```
+
+And
+
+```go
+type Request struct {
+  Method  string
+  Timeout int
+  secret  string
+  Con     *Connection
+}
+req := Request{Method: "GET", Timeout: 30, secret: "pony"}
+klog.InfoS("Request finished", "request", Request)
+```
+
+That would result in log (pretty printed)
+```json
+{
+   "ts": 1580306777.04728,
+   "v": 4,
+   "msg": "Request finished",
+   "request":{
+      "Method": "GET",
+      "Timeout": 30
+   }
+}
+```
+
+
+### Logging configuration
```

Comment:
This appears to be missing filtering per sink.  If we're going to introduce new logging configure APIs, that one seems like a big win.  The old glog vmodule has pretty significant limitations.

Reply:
What kind of filtering would you like to have?
The existing glog vmodule for loglevel will be still available.